### PR TITLE
Empty S3 bucket on deletion

### PR DESCRIPTION
**Pull request by sosodev**
If not protected ofc

### CODE REVIEW

File: utils/cdk-runner/pkg/aws/cloudformation/stack.go [hunk 1]
```diff
@@ -36,7 +36,7 @@ func GetStack(c *Client, stackName string) (*CfnStack, error) {
 		return cStack, err
 	}
 
-	if len(s.Stacks) > 0 {
+	if s != nil && len(s.Stacks) > 0 {
```
Comment (sosodev) [inline utils/cdk-runner/pkg/aws/cloudformation/stack.go:39]: We ignore the error generated above unless it's specifically the stack not existing. That means s can be nil here and cause a panic.

File: s3/Acornfile [hunk 2]
```diff
@@ -48,68 +50,94 @@ services: "admin": {
...
-	env: {
-		CDK_DEFAULT_ACCOUNT: "@{secrets.aws-context.account-id}"
-		CDK_DEFAULT_REGION:  "@{secrets.aws-context.aws-region}"
-		VPC_ID:              "@{secrets.aws-context.vpc-id}"
-		ACORN_ACCOUNT:       "@{acorn.account}"
-		ACORN_NAME:          "@{acorn.name}"
-		ACORN_PROJECT:       "@{acorn.project}"
-		ACORN_EXTERNAL_ID:   "@{acorn.externalID}"
+
+	apply: {
```
Comment (cloudnautique) [inline s3/Acornfile:77]: This should have a `dependsOn: ["cleanup"] ` ??

File: s3/Dockerfile [hunk 3]
```diff
@@ -8,7 +8,7 @@ RUN --mount=type=cache,target=/root/go/pkg \
     ls ../ &&\
     go build -o s3 .
 
-FROM ghcr.io/acorn-io/aws/utils/cdk-runner:v0.3.0 as cdk-runner
+FROM ghcr.io/acorn-io/aws/utils/cdk-runner:v0.4.0 as cdk-runner
```
Comment (cloudnautique) [inline s3/Dockerfile:11]: v0.5.0 is the latest now.

File: utils/cleanup/main.go [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+		if strings.Contains(err.Error(), "does not exist") {
+			// no cleanup is necessary
+			return
+		}
+
+		logrus.Fatal(err)
+	}
+
+	// check for deletion protection
+	if stack.DeletionProtection && os.Getenv(acornCf.DeletionProtectionEnvKey) == "true" {
```
Comment (cloudnautique) [inline utils/cleanup/main.go:49]: What if the stack has deletion protection and the user is trying to turn it off? thinking out loud, is this correct?
- The event is delete, so the user is trying to remove.
- deletion protection is turned off at the Acorn level, but has not been turned off on the AWS stack side.
- So we will empty the bucket now. 
- The apply job will run with delete protection off, and update the stack, then delete.

If that's correct, then I think we are good.

**Review by cloudnautique — CHANGES_REQUESTED** [review]
Should we call this bucket-cleanup or s3-cleanup? Since it is not a generic cleanup.

File: utils/cleanup/main.go (hunk 4 above)
Comment (sosodev) [reply utils/cleanup/main.go:49]: That seems correct. 👍

File: s3/Acornfile (hunk 2 above)
Comment (sosodev) [reply s3/Acornfile:77]: Good point. Added.

File: s3/Dockerfile (hunk 3 above)
Comment (sosodev) [reply s3/Dockerfile]: Bumped to 0.5.0

**Comment by sosodev** [comment]
> Should we call this bucket-cleanup or s3-cleanup? Since it is not a generic cleanup.

I've renamed it to s3-cleanup.

**Comment by sosodev** [comment]
Should fix https://github.com/acorn-io/manager/issues/1057

**Pull request merged**